### PR TITLE
[improve][txn] PIP-160: Pending ack log store enables the batch feature

### PR DESCRIPTION
**Pull request by poorbarcode**
Master Issue: #15370

### Motivation

see #15370

### Modifications

I will complete proposal #15370 with these pull requests( *current pull request is the step-4* ): 

1. Write the batch transaction log handler: `TxnLogBufferedWriter`
2. Configuration changes and protocol changes.
3. Transaction log store enables the batch feature.
4. Pending ack log store enables the batch feature.
5. Supports dynamic configuration.
6. Append admin API for transaction batch log and docs( admin and configuration doc ).
7. Append metrics support for transaction batch log.

### Documentation

- [ ] `doc-required` 
  
- [x] `doc-not-needed` 
  
- [ ] `doc` 

- [ ] `doc-complete`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 1]
```diff
@@ -240,13 +280,48 @@ public void addFailed(ManagedLedgerException exception, Object ctx) {
...
+            if (pendingAckLog.getPendingAckMetadatasList().isEmpty()){
+                continue;
+            }
+            for (PendingAckMetadata ack : pendingAckLog.getPendingAckMetadatasList()){
+                if (maxAcknowledgementPosition.compareTo(ack.getLedgerId(), ack.getEntryId()) < 0){
+                    maxAcknowledgementPosition = PositionImpl.get(ack.getLedgerId(), ack.getEntryId());
+                    pendingAckLogHasMaxAckPosition = pendingAckLog;
+                }
+            }
+        }
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:318]: only the logList.size() is batchSize, we handle once is enough. don't every add ops handle

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 2]
```diff
@@ -240,13 +280,48 @@ public void addFailed(ManagedLedgerException exception, Object ctx) {
...
+
+        if (pendingAckLogHasMaxAckPosition != null) {
+            handleMetadataEntry(logPosition, pendingAckLogHasMaxAckPosition);
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:321]: commit or abort all need handleMetadataEntry right?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 3]
```diff
@@ -104,6 +126,11 @@ public MLPendingAckStore(ManagedLedger managedLedger, ManagedCursor cursor,
         this.subManagedCursor = subManagedCursor;
         this.logIndexBackoff = new LogIndexLagBackoff(transactionPendingAckLogIndexMinLag, Long.MAX_VALUE, 1);
         this.maxIndexLag = logIndexBackoff.next(0);
+        this.bufferedWriter = new TxnLogBufferedWriter(managedLedger, ((ManagedLedgerImpl) managedLedger).getExecutor(),
+                scheduledExecutorService, PendingAckLogSerializer.INSTANCE,
+                bufferedWriterConfig.getBatchedWriteMaxRecords(), bufferedWriterConfig.getBatchedWriteMaxSize(),
+                bufferedWriterConfig.getBatchedWriteMaxDelayInMillis(), bufferedWriterConfig.isBatchEnabled());
+        this.batchedPendingAckLogsWaitingForHandle = new ArrayList<>(bufferedWriterConfig.getBatchedWriteMaxRecords());
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:133]: We don't need to set the size, we can wait for it to automatically expand

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 4]
```diff
@@ -240,13 +280,48 @@ public void addFailed(ManagedLedgerException exception, Object ctx) {
...
+            }
+            if (pendingAckLog.getPendingAckMetadatasList().isEmpty()){
+                continue;
+            }
+            for (PendingAckMetadata ack : pendingAckLog.getPendingAckMetadatasList()){
+                if (maxAcknowledgementPosition.compareTo(ack.getLedgerId(), ack.getEntryId()) < 0){
+                    maxAcknowledgementPosition = PositionImpl.get(ack.getLedgerId(), ack.getEntryId());
+                    pendingAckLogHasMaxAckPosition = pendingAckLog;
+                }
+            }
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:317]: this logic is very similar to 
https://github.com/apache/pulsar/blob/4483df2b3d113bfb55a8bb1a9e6c77160257c475/pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java#L255-L257

may we can handle once

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 5]
```diff
@@ -332,14 +408,41 @@ public void run() {
...
+                             * 1. Query batch index of current entry from cursor.
+                             * 2. Filter the data which has already ack.
+                             * 3. Build batched position and handle valid data.
+                             */
+                            long[] ackSetAlreadyAck = cursor.getDeletedBatchIndexesAsLongArray(
+                                    PositionImpl.get(entry.getLedgerId(), entry.getEntryId()));
+                            BitSetRecyclable bitSetAlreadyAck = null;
+                            if (ackSetAlreadyAck != null){
+                                bitSetAlreadyAck = BitSetRecyclable.valueOf(ackSetAlreadyAck);
+                            }
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:431]: pending ack clear log data is use markDelete, so we don't need this logic

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 6]
```diff
@@ -343,6 +346,8 @@ public BrokerService(PulsarService pulsar, EventLoopGroup eventLoopGroup) throws
         this.backlogQuotaManager = new BacklogQuotaManager(pulsar);
         this.backlogQuotaChecker = Executors
                 .newSingleThreadScheduledExecutor(new DefaultThreadFactory("pulsar-backlog-quota-checker"));
+        this.transactionLogBufferedWriteAsyncFlushTrigger = Executors
+                .newSingleThreadScheduledExecutor(new DefaultThreadFactory("pulsar-backlog-quota-checker"));
```
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:350]: Already fixed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 7]
```diff
@@ -240,13 +280,48 @@ public void addFailed(ManagedLedgerException exception, Object ctx) {
...
+     * @param logList Pending ack log records in a batch log Entry.
+     */
+    private void handleMetadataEntry(PositionImpl logPosition, List<PendingAckMetadataEntry> logList) {
+        // Find the record that carries the largest ack info, and call "handleMetadataEntry(position, pendingAckLog)"
+        PendingAckMetadataEntry pendingAckLogHasMaxAckPosition = null;
+        PositionImpl maxAcknowledgementPosition = this.maxAckPosition;
+        for (int i = logList.size() - 1; i >= 0; i--){
+            PendingAckMetadataEntry pendingAckLog = logList.get(i);
+            if (pendingAckLog.getPendingAckOp() == PendingAckOp.ABORT
+                    && pendingAckLog.getPendingAckOp() == PendingAckOp.COMMIT) {
```
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java]: Already fixed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java (hunk 4 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java]: Already fixed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java (hunk 5 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java]: Already fixed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java (hunk 3 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java]: Already fixed

**Comment by codelipenghui** [comment]
@poorbarcode #16727 has merged, could you please use the shared time for this PR?

**Comment by poorbarcode** [comment]
> @poorbarcode #16727 has merged, could you please use the shared time for this PR?

Already fixed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 8]
```diff
@@ -224,6 +224,9 @@ public class BrokerService implements Closeable {
     @Getter
     private final ScheduledExecutorService backlogQuotaChecker;
 
+    @Getter
+    private final ScheduledExecutorService transactionTimer;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:228]: Looks unused.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 9]
```diff
@@ -344,6 +347,8 @@ public BrokerService(PulsarService pulsar, EventLoopGroup eventLoopGroup) throws
         this.backlogQuotaManager = new BacklogQuotaManager(pulsar);
         this.backlogQuotaChecker = Executors
                 .newSingleThreadScheduledExecutor(new DefaultThreadFactory("pulsar-backlog-quota-checker"));
+        this.transactionTimer = Executors
+                .newSingleThreadScheduledExecutor(new DefaultThreadFactory("pulsar-backlog-quota-checker"));
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:351]: Looks unused.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 10]
```diff
@@ -362,6 +439,25 @@ public void run() {
...
 
+    private List<PendingAckMetadataEntry> deserializeEntry(Entry entry){
+        ByteBuf buffer = entry.getDataBuffer();
+        // Check whether it is batched Entry.
+        buffer.markReaderIndex();
+        short magicNum = buffer.readShort();
+        buffer.resetReaderIndex();
+        if (magicNum == BATCHED_ENTRY_DATA_PREFIX_MAGIC_NUMBER){
+            // skip version
+            buffer.skipBytes(4);
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:450]: Use the constant https://github.com/apache/pulsar/pull/16685/files#diff-4fb66809c520962f28616ddcab91aca0abb7aa04ee3ef0703097cf5976080ba8R301-R303

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStoreTest.java [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+        final AtomicInteger processController = new AtomicInteger();
+        doAnswer(new Answer() {
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+                processController.incrementAndGet();
+                return null;
+            }
+        }).when(pendingAckHandle).completeHandleFuture();
+        mlPendingAckStoreForRead.replayAsync(pendingAckHandle, internalPinnedExecutor);
+        Awaitility.await().atMost(200, TimeUnit.SECONDS).until(() -> processController.get() == 1);
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStoreTest.java:212]: Why need to wait at most 200 seconds? we don't have too many logs in the pending ack store.

**Review by codelipenghui — APPROVED** [review]
LGTM, left some minor comments.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 9 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java]: Already remove. Thanks.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java (hunk 10 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java]: Already fixed. Thanks.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStoreTest.java (hunk 11 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStoreTest.java]: Already change to 2 seconds.

**Review by congbobo184 — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**